File: Rectangle.Windows/src/Rectangle.Windows/Services/TodoManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Rectangle.Windows.Core;

namespace Rectangle.Windows.Services;

/// <summary>
/// Todo 模式管理器
/// 管理 Todo 应用窗口和其他窗口的协调布局
/// </summary>
public class TodoManager
{
    private readonly Win32WindowService _win32;
    private readonly ConfigService _configService;
    private readonly WindowTypeService _windowTypeService;

    public TodoManager(Win32WindowService win32, ConfigService configService, WindowTypeService windowTypeService)
    {
        _win32 = win32;
        _configService = configService;
        _windowTypeService = windowTypeService;
    }

    /// <summary>
    /// 检查 Todo 模式是否启用
    /// </summary>
    public bool IsTodoModeEnabled()
    {
        var config = _configService.Load();
        return config.TodoMode && !string.IsNullOrEmpty(config.TodoApplication);
    }

    /// <summary>
    /// 获取 Todo 应用窗口句柄
    /// </summary>
    public nint GetTodoWindow()
    {
        var config = _configService.Load();
        if (string.IsNullOrEmpty(config.TodoApplication))
            return 0;

        var windows = WindowEnumerator.EnumerateWindowsByProcess(config.TodoApplication);
        return windows.FirstOrDefault();
    }

    /// <summary>
    /// 调整 Todo 窗口到侧边栏位置
    /// </summary>
    public void AdjustTodoWindow(WorkArea workArea)
    {
        var config = _configService.Load();
        if (!config.TodoMode) return;

        var todoHwnd = GetTodoWindow();
        if (todoHwnd == 0) return;

        var sidebarWidth = config.TodoSidebarWidth;
        var isLeftSide = config.TodoSidebarSide == TodoSidebarSide.Left;

        int x = isLeftSide ? workArea.Left : workArea.Right - sidebarWidth;
        int y = workArea.Top;
        int width = sidebarWidth;
        int height = workArea.Height;

        _win32.SetWindowRect(todoHwnd, x, y, width, height);

        Logger.Info("TodoManager", $"调整 Todo 窗口到 {(isLeftSide ? "左侧" : "右侧")}: ({x}, {y}, {width}, {height})");
    }

    /// <summary>
    /// 获取排除 Todo 区域后的可用工作区
    /// </summary>
    public WorkArea GetAvailableWorkArea(WorkArea fullWorkArea)
    {
        var config = _configService.Load();
        if (!config.TodoMode || string.IsNullOrEmpty(config.TodoApplication))
            return fullWorkArea;

        var todoHwnd = GetTodoWindow();
        if (todoHwnd == 0) return fullWorkArea;

        var sidebarWidth = config.TodoSidebarWidth;
        var isLeftSide = config.TodoSidebarSide == TodoSidebarSide.Left;

        if (isLeftSide)
        {
            return new WorkArea(
                fullWorkArea.Left + sidebarWidth,
                fullWorkArea.Top,
                fullWorkArea.Right,
                fullWorkArea.Bottom
            );
        }
        else
        {
            return new WorkArea(
                fullWorkArea.Left,
                fullWorkArea.Top,
                fullWorkArea.Right - sidebarWidth,
                fullWorkArea.Bottom
            );
        }
    }

    /// <summary>
    /// 检查窗口是否是 Todo 应用窗口
    /// </summary>
    public bool IsTodoWindow(nint hwnd)
    {
        var config = _configService.Load();
        if (string.IsNullOrEmpty(config.TodoApplication))
            return false;

        var processName = WindowEnumerator.GetProcessNameFromWindow(hwnd);
        return processName.Equals(config.TodoApplication, StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// 获取所有非 Todo 窗口
    /// </summary>
    public List<nint> GetNonTodoWindows()
    {
        var allWindows = WindowEnumerator.EnumerateVisibleWindows();
        return allWindows.Where(hwnd => !IsTodoWindow(hwnd)).ToList();
    }

    /// <summary>
    /// 重新布局所有非 Todo 窗口（在可用区域内）
    /// </summary>
    public void RelayoutNonTodoWindows(WorkArea workArea)
    {
        var availableWorkArea = GetAvailableWorkArea(workArea);
        var windows = GetNonTodoWindows();

        if (windows.Count == 0) return;

        // 简单的平铺布局
        var tileManager = new TileAllManager(_win32, _configService);
        tileManager.TileAll(availableWorkArea);

        Logger.Info("TodoManager", $"重新布局 {windows.Count} 个非 Todo 窗口");
    }
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Rust/src/services/todo_manager.rs
use crate::core::rect::{WorkArea, WindowRect};
use crate::services::config::ConfigService;
use crate::services::win32::Win32WindowService;
use crate::services::window_enumerator::WindowEnumerator;
use crate::services::tile_all_manager::TileAllManager;
use crate::services::logger::Logger;

/// Todo 模式管理器
pub struct TodoManager {
    win32: Win32WindowService,
    config_service: ConfigService,
}

impl TodoManager {
    pub fn new(win32: Win32WindowService, config_service: ConfigService) -> Self {
        Self { win32, config_service }
    }

    /// 检查 Todo 模式是否启用
    pub fn is_todo_mode_enabled(&self) -> bool {
        let config = self.config_service.load();
        config.todo_mode && !config.todo_application.is_empty()
    }

    /// 获取 Todo 应用窗口句柄
    pub fn get_todo_window(&self) -> Option<isize> {
        let config = self.config_service.load();
        if config.todo_application.is_empty() {
            return None;
        }

        let windows = WindowEnumerator::enumerate_windows_by_process(&config.todo_application, true);
        windows.into_iter().next()
    }

    /// 调整 Todo 窗口到侧边栏位置
    pub fn adjust_todo_window(&self, work_area: &WorkArea) {
        let config = self.config_service.load();
        if !config.todo_mode {
            return;
        }

        let todo_hwnd = self.get_todo_window();
        if todo_hwnd.is_none() {
            return;
        }
        let todo_hwnd = todo_hwnd.unwrap();

        let sidebar_width = config.todo_sidebar_width;
        let is_left_side = config.todo_sidebar_side == 0; // 0 = Left

        let x = if is_left_side { work_area.left } else { work_area.right - sidebar_width };
        let y = work_area.top;
        let width = sidebar_width;
        let height = work_area.height();

        self.win32.set_window_rect(todo_hwnd, x, y, width, height);

        Logger::info("TodoManager", &format!("调整 Todo 窗口到 {}: ({}, {}, {}, {})",
            if is_left_side { "左侧" } else { "右侧" }, x, y, width, height));
    }

    /// 获取排除 Todo 区域后的可用工作区
    pub fn get_available_work_area(&self, full_work_area: &WorkArea) -> WorkArea {
        let config = self.config_service.load();
        if !config.todo_mode || config.todo_application.is_empty() {
            return *full_work_area;
        }

        let todo_hwnd = self.get_todo_window();
        if todo_hwnd.is_none() {
            return *full_work_area;
        }

        let sidebar_width = config.todo_sidebar_width;
        let is_left_side = config.todo_sidebar_side == 0;

        if is_left_side {
            WorkArea::new(
                full_work_area.left + sidebar_width,
                full_work_area.top,
                full_work_area.right,
                full_work_area.bottom,
            )
        } else {
            WorkArea::new(
                full_work_area.left,
                full_work_area.top,
                full_work_area.right - sidebar_width,
                full_work_area.bottom,
            )
        }
    }

    /// 检查窗口是否是 Todo 应用窗口
    pub fn is_todo_window(&self, hwnd: isize) -> bool {
        let config = self.config_service.load();
        if config.todo_application.is_empty() {
            return false;
        }

        let process_name = WindowEnumerator::get_process_name_from_window(hwnd);
        process_name.eq_ignore_ascii_case(&config.todo_application)
    }

    /// 获取所有非 Todo 窗口
    pub fn get_non_todo_windows(&self) -> Vec<isize> {
        let all_windows = WindowEnumerator::enumerate_visible_windows(true, true);
        all_windows.into_iter()
            .filter(|hwnd| !self.is_todo_window(*hwnd))
            .collect()
    }

    /// 重新布局所有非 Todo 窗口（在可用区域内）
    pub fn relayout_non_todo_windows(&self, work_area: &WorkArea) {
        let available_work_area = self.get_available_work_area(work_area);
        let windows = self.get_non_todo_windows();

        if windows.is_empty() {
            return;
        }

        // 使用平铺布局
        let tile_manager = TileAllManager::new(self.win32.clone(), Some(self.config_service.clone()));
        // 平铺窗口
        for (index, hwnd) in windows.iter().enumerate() {
            if index >= windows.len() {
                break;
            }
            // 简化的平铺逻辑
            let col = (index % 2) as i32;
            let row = (index / 2) as i32;
            let cols = 2;
            let rows = (windows.len() + 1) / 2;

            let cell_width = available_work_area.width() / cols as i32;
            let cell_height = available_work_area.height() / rows as i32;

            let x = available_work_area.left + col * cell_width;
            let y = available_work_area.top + row * cell_height;

            self.win32.set_window_rect(*hwnd, x, y, cell_width, cell_height);
        }

        Logger::info("TodoManager", &format!("重新布局 {} 个非 Todo 窗口", windows.len()));
    }

    /// 切换 Todo 模式
    pub fn toggle_todo_mode(&self, work_area: &WorkArea) {
        if self.is_todo_mode_enabled() {
            self.adjust_todo_window(work_area);
            self.relayout_non_todo_windows(work_area);
        }
    }
}
